File: Python3/1852. Distinct Numbers in Each Subarray.py
class Solution:
    def distinctNumbers(self, nums: List[int], k: int) -> List[int]:
        num_count = defaultdict(int)
        distinct_count = 0
        ans = []

        for i, n in enumerate(nums):
            if num_count[n] == 0:
                distinct_count += 1
            num_count[n] += 1

            if i >= k:
                num_count[nums[i - k]] -= 1
                if num_count[nums[i - k]] == 0:
                    distinct_count -= 1

            if i >= k - 1:
                ans.append(distinct_count)

        return ans

File: Java/1852. Distinct Numbers in Each Subarray.java
class Solution {
    public int[] distinctNumbers(int[] nums, int k) {
        Map<Integer, Integer> num_count = new HashMap<>();
        int distinct_count = 0, N = nums.length;
        int[] ans = new int[N - k + 1];

        for (int i = 0; i < N; i++) {
            if (num_count.getOrDefault(nums[i], 0) == 0) {
                distinct_count++;
            }
            num_count.put(nums[i], num_count.getOrDefault(nums[i], 0) + 1);
            if (i >= k) {
                num_count.put(nums[i - k], num_count.get(nums[i - k]) - 1);
                if (num_count.get(nums[i - k]) == 0) {
                    distinct_count--;
                }
            }

            if (i >= k - 1) {
                ans[i - k + 1] = distinct_count;
            }
        }
        return ans;
    }
}